torch/utils/data/dataset.py

```diff
@@ -260,6 +265,132 @@ def __len__(self):
+    data into the ``ChunkDataset`` internal cache. Only then ``ChunkDataset`` returns batches.
+
+    Arguments:
+        chunk_sampler (DistributedSampler or DistributedChunkSampler): Draw indices for chunks.
+            Typically used to split data amongst dataloader workers (first level of sampling)
+        chunk_reader (ChunkDataReader): Specialized reader for a given input type
+        shuffle_cache (bool): Setting `True` trigger shuffling of the internal chunk cache
+            (second level of sampling) (default: `True`)
+    """
+
+    def __init__(self, chunk_sampler, chunk_reader, shuffle_cache=True):
+        super(ChunkDataset, self).__init__()
+        assert isinstance(chunk_sampler, DistributedSampler) or isinstance(
+            chunk_sampler, DistributedChunkSampler), 'sampler must be a `DistributedSampler` or `DistributedChunkSampler`'
+        assert callable(chunk_reader), 'chunk_reader must be `callable()` and return a container with data'
+        assert isinstance(shuffle_cache, bool), 'shuffle_cache must be a `bool`'
+
+        self.chunk_sampler = chunk_sampler
+        self.chunk_reader = chunk_reader
+        self.shuffle_cache = shuffle_cache
+
+        # Internal state
+        self._chunk_sampler_iter = None
+        self._cache = []
+        self._min_cache = 1000
+        self._is_ready = False
+
+    def __iter__(self):
+        return self
+
+    def __len__(self):
+        # `IterableDataset` classes have unknown dataset size
+        raise NotImplementedError
+
+    def __next__(self):
+        r"""Returns a batch or raises exception when exhausted"""
+
+        # Run once: update global worker rank on each worker process
+        if not self._is_ready:
+            self._update_global_worker_rank()
+
+        cache_size = 0
+        if len(self._cache) > 0:
+            cache_size = len(self._cache)
+
+        if (cache_size < self._min_cache):
+            new_chunk = None
+            # Try reading a new chunk
+            # Ignore EOF here because there may be internal cache entries to be returned yet
+            try:
+                new_chunk = self.chunk_reader(next(self._chunk_sampler_iter))
+                assert isinstance(new_chunk, list), "Chunk data reader must return a `list(samples)`"
+            except StopIteration:
+                pass
+
+            # Apply second level of shuffling
+            if new_chunk and self.shuffle_cache:
+                random.shuffle(new_chunk)
+
+            # Updates internal cache
+            if new_chunk and cache_size > 0:
+                self._cache = self._cache + new_chunk
+            elif new_chunk:
+                self._cache = new_chunk
+            elif not new_chunk and cache_size == 0:
+                raise StopIteration
+
+        # Get a batch and update internal cache cache
+        if len(self._cache) > 0:
+            batch = self._cache[:1]
```

Comment:
* Why is this taking the first element as a list and then returning the element itself?
* What happens when the condition is not met and batch is not defined at the return statement?

Reply:
- Slicing the list was the simplest way to remove the first element from buffer and assign it to another variable in a single shot. Also, because `default_collate` creates a new list appending elements from several `next()` calls, we strip the existing list to avoid a double list (e.g. a single element list such as [1] with batch_size=4 would become [[1,2,3,4]] instead of [1,2,3,4]).
- If cache is empty, an exception is raised on line 352 and this code would never execute

Reply:
Couldn't you just use `batch = self._cache.pop(0)` to avoid the need for repeat slicing and indexing?

Reply:
Definitely. The first versions had the option of returning more than one element, so we used slice for it. The final design, however, ended up returning a single element, so `batch = self._cache.pop(0)` is nice. Thanks!